File: ref/bitunix-demo/Demo/Golang/open_api_http_sign.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/google/uuid"
)

// GetNonce generates a random string as nonce
func GetNonce() string {
	// generate UUID and remove hyphens
	return strings.ReplaceAll(uuid.New().String(), "-", "")
}

// GetTimestamp gets current timestamp in milliseconds
func GetTimestamp() string {
	return fmt.Sprintf("%d", time.Now().UnixMilli())
}

// GenerateSignature generates signature according to Bitunix OpenAPI doc
func GenerateSignature(apiKey, secretKey, nonce, timestamp, queryParams, body string) string {
	digestInput := strings.TrimSpace(nonce + timestamp + apiKey + queryParams + body)
	digest := sha256.Sum256([]byte(digestInput))
	digestStr := hex.EncodeToString(digest[:])

	signInput := strings.TrimSpace(digestStr + secretKey)
	sign := sha256.Sum256([]byte(signInput))
	return hex.EncodeToString(sign[:])
}

// GetAuthHeaders gets authentication headers
func GetAuthHeaders(apiKey, secretKey, queryParams, body string) map[string]string {
	nonce := GetNonce()
	timestamp := GetTimestamp()

	sign := GenerateSignature(
		apiKey,
		secretKey,
		nonce,
		timestamp,
		queryParams,
		body,
	)

	return map[string]string{
		"api-key":   apiKey,
		"sign":      sign,
		"nonce":     nonce,
		"timestamp": timestamp,
	}
}

// SortParams sorts parameters and concatenates them
func SortParams(params map[string]string) string {
	if len(params) == 0 {
		return ""
	}

	// get all keys and sort them
	keys := make([]string, 0, len(params))
	for k := range params {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	// concatenate parameters by sorted keys
	var builder strings.Builder
	for _, k := range keys {
		builder.WriteString(k)
		builder.WriteString(params[k])
	}
	return builder.String()
}

File: ref/bitunix-demo/Demo/Node/openApiHttpSign.js
const crypto = require('crypto');

/**
 * OpenAPI HTTP signature tool class
 */
class OpenApiHttpSign {
    /**
     * generate a random string as nonce
     * @returns {string} 32-bit random string
     */
    static getNonce() {
        return crypto.randomBytes(16).toString('hex');
    }

    /**
     * get current timestamp (millisecond)
     * @returns {string} millisecond timestamp
     */
    static getTimestamp() {
        return Date.now().toString();
    }

    /**
     * generate signature
     * @param {string} apiKey API key
     * @param {string} secretKey secret key
     * @param {string} nonce random string
     * @param {string} timestamp timestamp
     * @param {string} queryParams sorted query params string (no spaces)
     * @param {string} body original JSON string (no spaces)
     * @returns {string} signature
     */
    static generateSignature(
        apiKey,
        secretKey,
        nonce,
        timestamp,
        queryParams = "",
        body = ""
    ) {
        const digestInput = nonce + timestamp + apiKey + queryParams + body;
        const digest = crypto.createHash('sha256').update(digestInput).digest('hex');
        const signInput = digest + secretKey;
        return crypto.createHash('sha256').update(signInput).digest('hex');
    }

    /**
     * get authentication header information
     * @param {string} apiKey API key
     * @param {string} secretKey secret key
     * @param {Object} queryParams query parameters
     * @param {string} body request body
     * @returns {Object} authentication header information
     */
    static getAuthHeaders(
        apiKey,
        secretKey,
        queryParams = {},
        body = ""
    ) {
        const nonce = this.getNonce();
        const timestamp = this.getTimestamp();
        const queryParamsStr = this.sortParams(queryParams);

        const sign = this.generateSignature(
            apiKey,
            secretKey,
            nonce,
            timestamp,
            queryParamsStr,
            body
        );
        
        return {
            "api-key": apiKey,
            "sign": sign,
            "nonce": nonce,
            "timestamp": timestamp
        };
    }

    /**
     * sort parameters and concatenate
     * @param {Object} params parameter dictionary
     * @returns {string} sorted parameter string
     */
    static sortParams(params) {
        if (!params || Object.keys(params).length === 0) {
            return "";
        }
        
        // sort by key and concatenate directly
        return Object.keys(params)
            .sort()
            .map(key => key + params[key])
            .join('');
    }
}

module.exports = OpenApiHttpSign; 